### PR TITLE
feat: beta

### PR DESCRIPTION
**Pull request by tusharpandey13**
## Overview

Official Auth0 SDK for the Hono framework. One middleware call adds login, logout, session management, token access, and route protection. 
Works across Node.js, Cloudflare Workers, Bun, Deno, and Vercel Edge. Built on `@auth0/auth0-server-js`.

## Changes

- Zero-config `auth0()` middleware with env-based configuration
- Eager session/user/org loading via `c.var.auth0`
- Authorization middleware: `requiresAuth`, `requiresOrg`, `claimEquals`, `claimIncludes`, `claimCheck`
- Token helpers: `getAccessToken`, `getAccessTokenForConnection`, `updateSession` with request-level dedup
- Typed error hierarchy: `Auth0Error` + 7 subclasses extending Hono's `HTTPException`
- `onCallback` lifecycle hook for session enrichment
- Standalone route handlers: `handleLogin`, `handleLogout`, `handleCallback`, `handleBackchannelLogout`
- `HonoCookieHandler` with ALS fallback for multi-runtime session handling
- Security hardening: crash prevention, data integrity guards, Zod-based state validation

Please take a look at [Review guide](https://github.com/auth0/auth0-hono/pull/19#issuecomment-4294328847) for detailed description of changes and design decisions.

## Tests

325 tests across 20 files (751ms). Includes contract-enforcing mocks, integration tests with real `StatelessStateStore`, and live smoke testing against Auth0 tenant.

## Usage

```typescript
import { Hono } from 'hono'
import { auth0, requiresAuth, getAccessToken } from '@auth0/auth0-hono'

const app = new Hono()
app.use('*', auth0())
app.use('/dashboard/*', requiresAuth())

app.get('/profile', (c) => c.json(c.var.auth0.user))

app.get('/api/data', requiresAuth(), async (c) => {
const { accessToken } = await getAccessToken(c)
// call your API
})

### CODE REVIEW

**Comment by tusharpandey13** [comment]
## Review Guide

### What this PR is

Beta release of `@auth0/auth0-hono` : a full-featured Auth0 SDK for Hono with feature parity to `nextjs-auth0` and `auth0-fastify`.

### Architecture

Built on `@auth0/auth0-server-js` (handles OIDC, tokens, crypto). This SDK provides the Hono integration layer: middleware, context population, cookie handling, and developer API.

Key files by concern:
- Entry point: `src/auth.ts` : auth0() middleware, lazy client init, eager session loading, route mounting
- Config: `src/config/Schema.ts` (Zod validation), Configuration.ts (types), index.ts (cache + client factory)
- Session: `src/helpers/session.ts` (getSession, updateSession, `StateData` validation), src/session/`HonoCookieHandler`.ts (cookie I/O with ALS)
- Middleware: `src/middleware/` : login, callback, logout, backchannelLogout, silentLogin, requiresAuth, requiresOrg, claims
- Errors: `src/errors/Auth0Error.ts` (base), errorMap.ts (server-js → SDK error mapping)
- Types: `src/types/auth0.ts` (`Auth0Context`, `OIDCEnv`), `src/lib/honoEnv.ts` (`ContextVariableMap` augmentation)

### How to review

#### Phase 1 : Core flow (start here):
1. `src/auth.ts` : middleware chain, client init, session loading, route dispatch
2. `src/middleware/callback.ts` : OIDC callback, hook enrichment, internal field preservation
3. `src/helpers/session.ts` : `StateData` validation (Zod), updateSession with stateStore access

#### Phase 2 : Config & session handling:
4. `src/config/Schema.ts` : Zod schema with immutable .transform() for defaults
5. `src/session/`HonoCookieHandler`.ts` : cookie parsing with decode fallback, `Auth0Error` for ALS

#### Phase 3 : Middleware & helpers:
6. `src/middleware/login.ts, logout.ts, silentLogin.ts` : standard flows
7. `src/middleware/claims.ts, requiresOrg.ts, requiresAuth.ts` : authorization
8. `src/helpers/getAccessToken.ts` : token refresh with dedup cache

#### Phase 4 : Tests:
9. `test/mocks/stateStore.mock.ts` : contract-enforcing mock factory
10. Flow tests in `test/flows/` : end-to-end middleware chain tests
11. Unit tests in `test/helpers/, test/middleware/` : per-component coverage

### Key design decisions

- `SessionData` vs `StateData`: `getSession()` strips internal (server metadata). Code that persists must use `tateStore.get()` directly to preserve internal.createdAt.
- Eager loading: `c.var.auth0` populated on every request (not lazy). Trade-off: one extra decrypt per request, but zero-ceremony access.
- Error hierarchy: All SDK errors extend `Auth0Error` which extends Hono's `HTTPException`. Caught by Hono's `app.onError`.
- Multi-runtime: `HonoCookieHandler` uses `AsyncLocalStorage` with explicit context fallback. No process.env : all config via Hono's env(c).

### Security hardening

The last commit adds systematic fixes for 14 unsafe assumptions found via code audit:
- Crash prevention: try/catch on decodeURIComponent, parseBody, array query params
- Data integrity: internal field preservation, Zod-based `StateData` validation, config mutation prevention
- Type safety: `Auth0Error` consistency, instanceof Map checks, JSON cache keys
- Ordering: silent login cookie recovery, error-path cache invalidation

**Comment by tusharpandey13** [comment]
## Testing App
 
Live smoke test app exercising SDK's public API surface against a real Auth0 tenant.

**Repo:** https://github.com/tusharpandey13/auth0-hono-testing-app
 
**What it covers:** 
- Session update resilience (original `internal.createdAt` bug regression guard) 
- Malformed cookie handling
- Hook enrichment via `onCallback`
- Authorization middleware: `claimEquals`, `claimCheck`, `requiresOrg`
- Token helpers: `getAccessToken` dedup, `getAccessTokenForConnection`
- Concurrent `updateSession` (last-write-wins documentation)
- Edge cases: session state after logout intent 

**Recording:** [`test-recording.json`](https://github.com/tusharpandey13/auth0-hono-testing-app/blob/main/test-recording.json) — auto-generated on each test run, secrets masked. 
 
**Results (latest run):** 9 tests — 7 pass, 2 skip (expected: hook enrichment requires re-login, connection token requires google-oauth2 config), 0 fail.
 
**To reproduce:** 
```bash
git clone https://github.com/tusharpandey13/auth0-hono-testing-app
# fill Auth0 credentials
cp .env.example .env
npm install && npm run dev
# Login at http://localhost:3000, browse test routes
# Recording auto-saves to test-recording.json
```

File: src/helpers/getAccessToken.ts [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+  // Check if another handler is already refreshing this audience
+  let promise = refreshCache.get(cacheKey)
+  if (!promise) {
+    // No in-flight refresh — create new promise
+    // server-js client.getAccessToken() handles:
+    // - Checking token expiry
+    // - Refreshing if needed
+    // - Persisting new token to state store
+    // - Returning full TokenSet (not just string)
+    promise = client.getAccessToken(c)
```
Comment (arpit-jn) [inline src/helpers/getAccessToken.ts]: `getAccessToken` ignores audience parameter. `options?.audience` is used for cache key dedup but never forwarded to `client.getAccessToken(c)`. Users who pass `{ audience: 'https://api.example.com' }` will silently get the default audience token.

File: src/types/auth0.ts [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+ * @example
+ * ```typescript
+ * const org = c.var.auth0.org
+ * if (org) {
+ *   console.log(org.id)   // 'org_abc123'
+ *   console.log(org.name) // 'Acme Corp'
+ * }
+ * ```
+ */
+export interface Auth0Organization {
```
Comment (arpit-jn) [inline src/types/auth0.ts:54]: Singleton init captures first request's env. `initPromise` caches config from the first request's `env(c)`. Document this as a known limitation or add env-drift detection for multi-tenant serverless.

**Review by arpit-jn — CHANGES_REQUESTED** [review]

File: src/middleware/callback.ts [hunk 3]
```diff
@@ -1,53 +1,163 @@
...
+              configuration.debug(
+                'Warning: Hook enrichment discarded due to missing session state after successful login. ' +
+                'This may indicate a race condition in your state store.'
+              );
+            }
+          }
+          // void/undefined: use default behavior
+        } catch (hookErr) {
+          // Hook threw — log but don't mask the login
+          console.error('onCallback hook error:', hookErr)
```
Comment (arpit-jn) [inline src/middleware/callback.ts]: Hook errors use `console.error` instead of `configuration.debug()`. inconsistent with SDK logging pattern; may be invisible in serverless environments

File: src/config/index.ts [hunk 4]
```diff
@@ -1,33 +1,80 @@
...
+import { Context } from 'hono'
+import { Auth0Error } from '@/errors/index.js'
+import { initializeOidcClient } from '@/lib/client.js'
+import { STATE_STORE_KEY } from '@/lib/constants.js'
+import { Configuration, InitConfiguration } from './Configuration.js'
+import { ConfigurationSchema } from './Schema.js'
+import { assignFromEnv } from './envConfig.js'
 
-const parsedConfig = new Map<InitConfiguration, Configuration>();
+const parsedConfig = new Map<InitConfiguration, Configuration>()
```
Comment (arpit-jn) [inline src/config/index.ts]: `parsedConfig` Map keyed by object reference, standalone handlers via `ensureClient()` create new objects per call, so the config cache never hits

File: test/helpers/session.test.ts [hunk 5]
```diff
@@ -0,0 +1,754 @@
...
+
+    it('should handle session without internal field', async () => {
+      const sessionWithoutInternal = {
+        user: { sub: 'user123' },
+        idToken: 'token',
+      }
+
+      // REQ-B1: persistSession must reject sessions without internal field
+      // to prevent loss of critical internal state (sid, createdAt, etc)
+      expect(persistSession(mockContext, sessionWithoutInternal)).rejects.toThrow(
```
Comment (arpit-jn) [inline test/helpers/session.test.ts]: Missing `await` in test assertion. Will become a test failure in Vitest 3

File: src/types/auth0.ts (hunk 2 above)
Comment (tusharpandey13) [reply src/types/auth0.ts:54]: This is by-design for Beta, same pattern as nextjs-auth0 and auth0-fastify (single-app middleware). Added inline code comment documenting the known limitation. Multi-tenant serverless (different env bindings per route/request) is post-Beta scope. Env-drift detection is a good idea for that future work.

File: src/helpers/getAccessToken.ts (hunk 1 above)
Comment (tusharpandey13) [reply src/helpers/getAccessToken.ts]: Fixed: removed audience from `GetAccessTokenOptions` entirely. For connection-scoped tokens (e.g., Google OAuth2), `getAccessTokenForConnection()` is the correct API.

**Review by arpit-jn — DISMISSED** [review]

File: test/helpers/session.test.ts (hunk 5 above)
Comment (tusharpandey13) [reply test/helpers/session.test.ts]: fixed

File: src/config/index.ts (hunk 4 above)
Comment (tusharpandey13) [reply src/config/index.ts]: fixed

File: src/middleware/callback.ts (hunk 3 above)
Comment (tusharpandey13) [reply src/middleware/callback.ts]: fixed

File: src/types/session.ts
Comment (Piyush-85) [inline src/types/session.ts]: The JSDoc default here is wrong. The actual default is appSession, set in src/config/Schema.ts:39 and explicitly passed as stateIdentifier in src/lib/client.ts:103. The __a0_session string is the fallback inside auth0-server-js's ServerClient constructor — it's never reachable from this SDK because client.ts always passes a value. Should be Default: `appSession`.

File: src/middleware/requiresAuth.ts
Comment (Piyush-85) [inline src/middleware/requiresAuth.ts]: auth0() already calls getCachedSession(c) before this middleware runs and stores the result in context. Calling client.getSession(c) here does a redundant cookie parse + JWE decrypt on every protected route. Should use getCachedSession(c).

File: src/middleware/callback.ts [hunk 6]
```diff
@@ -1,53 +1,163 @@
...
+            return hookResult;
+          }
+          // Per design M4: any other return on error path is ignored
+        } catch {
+          // Hook error silently ignored — original auth error always propagates
+        }
+      }
+
+      // Always throw original error — hook failure never masks it
       await resumeSilentLogin()(c, next);
```
Comment (Piyush-85) [inline src/middleware/callback.ts]: **resumeSilentLogin() on error path can cause redirect loop**

On the success path, deleting the silent-login skip cookie is correct. On the error path, if the failure was from a prompt=none silent login, clearing this cookie means the next request re-triggers silent login → [/auth/callback] error → clear cookie → loop. Should call cancelSilentLogin() (sets the cookie) or nothing on the error path.

File: src/session/HonoCookieHandler.ts [hunk 7]
```diff
@@ -16,48 +17,73 @@ export class HonoCookieHandler implements CookieHandler<any> {
...
+    const ctx = this.getContext(storeOptions);
     setCookie(ctx, name, value, cookieOptions);
     return value;
   }
 
-  getCookie(name: string): string | undefined {
-    const ctx = HonoCookieHandler.getContext();
+  getCookie(name: string, storeOptions?: Context): string | undefined {
+    const ctx = this.getContext(storeOptions);
     return getCookie(ctx, name);
```
Comment (Piyush-85) [inline src/session/HonoCookieHandler.ts]: ## malformed session cookie crashes the entire middleware with a 500

The `getCookies()` method (line 46) has a `try/catch` around `decodeURIComponent` and its JSDoc explicitly claims it protects against "attacker-injected malformed cookies". However, this protection is on the **wrong method** and does not guard the actual session read path.

### How `auth0-server-js` reads the session

`StatelessStateStore.get()` does the following:
1. Calls `getCookieKeys()` → which calls `getCookies()` → only to **enumerate cookie names**
2. Calls **`getCookie(key)`** for each chunk → to **read the actual session value**

So `getCookies()` never sees the session cookie _value_ — only its name. The actual value decoding happens in `getCookie()` at line 82, which delegates to Hono's `getCookie(ctx, name)`.

### Why this crashes

Hono 4.7.x (the version in the SDK's own `node_modules`) calls `decodeURIComponent` **without a try-catch** in its cookie parser. Sending a request with: Cookie: appSession=%GG

**Fix:**
```ts
getCookie(name: string, storeOptions?: Context): string | undefined {
  const ctx = this.getContext(storeOptions);
  try {
    return getCookie(ctx, name);
  } catch {
    return undefined; // malformed %-encoding — treat as no session
  }
}

File: README.md [hunk 8]
```diff
@@ -1,335 +1,542 @@
...
-```ts
-import { Hono } from "hono";
-import { auth } from "@auth0/auth0-hono";
+| Variable | Required | Description |
+|----------|----------|-------------|
+| `AUTH0_DOMAIN` | Yes | Auth0 domain (e.g., `tenant.auth0.com`) |
+| `AUTH0_CLIENT_ID` | Yes | Auth0 application client ID |
+| `AUTH0_CLIENT_SECRET` | No | Client secret (required for refresh token flow) |
+| `AUTH0_SESSION_ENCRYPTION_KEY` | Yes | 32+ character encryption key for session cookies |
+| `APP_BASE_URL` | Yes | Base URL of your application (e.g., `https://myapp.com`) |
```
Comment (Piyush-85) [inline README.md:55]: The README documents APP_BASE_URL but src/config/envConfig.ts:6 defines the key as [BASE_URL](vscode-file://vscode-app/Applications/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-browser/workbench/workbench.html). Any developer following the quickstart gets a Zod validation error on startup ([baseURL: Required](vscode-file://vscode-app/Applications/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-browser/workbench/workbench.html)). 

This can be standardised to be what other auth0 SDks use.

File: src/utils/util.ts [hunk 9]
```diff
@@ -25,14 +23,14 @@ export function toSearchParams(
 export function validateRedirectUrl(url: string, baseURL?: string): string {
```
Comment (Piyush-85) [inline src/utils/util.ts]: This appears to be dead code

File: src/config/Schema.ts [hunk 10]
```diff
@@ -63,123 +60,114 @@ export const ConfigurationSchema = z
...
     httpTimeout: z.number().min(500).optional().default(5000),
-    httpUserAgent: z.string().optional().default("hono-openid-connect"),
+    httpUserAgent: z.string().optional().default('hono-openid-connect'),
 
     fetch: z
-      .custom<typeof globalThis.fetch>((v) => typeof v === "function")
+      .custom<typeof globalThis.fetch>((v) => typeof v === 'function')
       .optional()
       .default(() => globalThis.fetch),
   })
```
Comment (Piyush-85) [inline src/config/Schema.ts:119]: **`onCallback` silently stripped by Zod, hook never fires via `auth0()`**

`onCallback` is not in `ConfigurationSchema`. Zod strips unknown keys by default, so passing it to `auth0()` is silently discarded at parse time.

In `callback.ts:70` and `callback.ts:129`:

```ts
const hook = params.onCallback ?? configuration.onCallback;
// → undefined ?? undefined → hook is never called
```

The hook only works via the standalone handler:

```ts
handleCallback({ onCallback: myHook })  // ✅ works
auth0({ onCallback: myHook })           // ❌ silently ignored
```

The README documents `auth0({ onCallback })` as the primary usage — a developer following it will see no error and no effect.

**Fix** — add before the closing `})` on this line:

```ts
onCallback: z
  .custom<Configuration['onCallback']>((v) => typeof v === 'function')
  .optional(),
```

File: src/helpers/session.ts [hunk 11]
```diff
@@ -0,0 +1,227 @@
...
+import { SessionData, StateData, StateStore } from '@auth0/auth0-server-js';
+import { RESERVED_FIELDS, STATE_STORE_KEY } from '@/lib/constants.js';
+import { Auth0Session, Auth0User } from '@/types/auth0.js';
+import { MissingSessionError } from '@/errors/index.js';
+import { Auth0Error } from '@/errors/Auth0Error.js';
+
+const StateDataSchema = z
+  .object({
+    user: z.record(z.any()),
+    idToken: z.string(),
```
Comment (Piyush-85) [inline src/helpers/session.ts]: **`idToken` required in schema but `string | undefined` in `auth0-server-js`**

```ts
idToken: z.string(),  // ← required
```

`auth0-server-js` types `StateData.idToken` as `string | undefined`. For any flow that doesn't issue an ID token (pure M2M, `response_type: token`), every session write throws a Zod validation error and the session is never persisted.

File: src/lib/Exception.ts [hunk 12]
```diff
@@ -1,12 +1,12 @@
-import { HTTPException } from "hono/http-exception";
-import { ContentfulStatusCode } from "hono/utils/http-status";
+import { HTTPException } from 'hono/http-exception';
+import { ContentfulStatusCode } from 'hono/utils/http-status';
 
 export class Auth0Exception extends HTTPException {
```
Comment (Piyush-85) [inline src/lib/Exception.ts]: **orphaned class, creates a confusing duplicate of `Auth0Exception`**

This file defines `Auth0Exception` extending `HTTPException` directly with a different constructor signature. It is never imported anywhere in `src/`.

The `Auth0Exception` that is actually exported from the package (`src/index.ts:42`) is:

```ts
// src/errors/index.ts
export { Auth0Error as Auth0Exception };
```

Two things named `Auth0Exception` with incompatible shapes will confuse anyone reading the source. `src/lib/Exception.ts` should be deleted before release.

File: src/config/Schema.ts [hunk 13]
```diff
@@ -27,7 +24,7 @@ export const ConfigurationSchema = z
     sessionStore: z.instanceof(SessionStore).optional(),
```
Comment (Piyush-85) [inline src/config/Schema.ts]: **top-level `sessionStore` key is accepted by Zod but never read — custom store silently has no effect**

`ConfigurationSchema` declares `sessionStore` as a valid top-level option:

```ts
sessionStore: z.instanceof(SessionStore).optional(),
```

Zod accepts it without error. But `client.ts` only reads `config.session.store` when choosing between `StatefulStateStore` and `StatelessStateStore`:

```ts
const baseStore = config.session.store
  ? new StatefulStateStore(...)
  : new StatelessStateStore(...);   // ← always chosen when sessionStore is used
```

`config.sessionStore` has zero usages anywhere in `src/`. A developer passing:

```ts
auth0({ sessionStore: myStore })  // ❌ silently ignored
```

gets no error, no warning, and no custom store — they get stateless cookie sessions. Their store's `set`, `get`, and `delete` methods are never called. The working API is:

```ts
auth0({ session: { store: myStore } })  // ✅
```

**Fix:** remove `sessionStore` from the schema entirely (it has no corresponding consumer), or wire it through to `session.store` in the `transform` block.

File: src/types/session.ts
Comment (Piyush-85) [inline src/types/session.ts]: **`secret` typed as `string` but schema and README both support `string | string[]`**

**Comment by Piyush-85** [comment]
@tusharpandey13 Found two other potential bugs while testing, posting them below PTAL

### [BUG-1] `cancelSilentLogin()` never calls `next()` — silently breaks logout composition
- **File:** `src/middleware/silentLogin.ts:32`
- **Repro:** The most natural real-world usage:
  ```ts
  app.get('/logout', cancelSilentLogin(), handleLogout({ idpLogout: true }))
  ```
  `cancelSilentLogin()` sets the cookie but never calls `next()`, so `handleLogout()` never runs. The user gets a null-body response and is **not logged out**. In the Node.js adapter this also throws `Error: res.body.getReader is not a function`. The same issue affects any handler chained after `cancelSilentLogin()`:
  ```ts
  app.get('/signed-out', cancelSilentLogin(), (c) => c.html('<h1>Bye!</h1>'))
  app.get('/opt-out',    cancelSilentLogin(), (c) => c.redirect('/'))
  ```
- **Root cause:** `createMiddleware` callback in `cancelSilentLogin()` calls `setCookie()` then returns implicitly without calling `await next()`, unconditionally terminating the middleware chain.
- **Fix:** Add `return next()` after `setCookie()`.
  ```ts
  export const cancelSilentLogin = () =>
    createMiddleware(async (c, next) => {   // <-- add `next` param
      setCookie(c, COOKIE_NAME, 'true', getCookieOptions(c));
      return next();                         // <-- add this
    });
  ```
  
### [BUG-2] `onCallback` hook always receives `null` session — enrichment silently discarded
- **File:** `src/middleware/callback.ts:68` + `src/session/HonoCookieHandler.ts:47`
- **Repro:** Register an `onCallback` hook via `handleCallback({ onCallback })`. After successful login, the hook fires but `session` parameter is always `null`. Any enrichment returned by the hook is silently discarded.
- **Root cause:** In `callback.ts`, after `completeInteractiveLogin()` the code immediately calls `client.getSession(c)` to get the session to pass to the hook (line 68). But `completeInteractiveLogin()` stores the session via `setCookie()` which writes to the **response headers**. `getSession()` reads cookies via `getCookies()` which reads from `req.header('Cookie')` — the **incoming request**. The newly set cookie does not exist in the request, so `getSession()` always returns `null` on the callback request. Then at line 88, `stateStore.get()` has the same problem — returns `null` — so the enrichment merge is never executed and the code hits the "race condition" warning path and discards enrichment silently.
- **Fix:** `completeInteractiveLogin()` should return the `StateData`/`SessionData` that was just stored, so `callback.ts` can pass it directly to the hook without re-reading from the cookie store. Alternatively, `HonoCookieHandler` needs a write-through cache so cookies set in the current request are readable within the same request.

File: src/types/session.ts
Comment (tusharpandey13) [reply src/types/session.ts]: Fixed: changed to `Default: appSession`.

File: src/middleware/requiresAuth.ts
Comment (tusharpandey13) [reply src/middleware/requiresAuth.ts]: Fixed: uses getCachedSession(c) instead of client.getSession(c).

File: src/middleware/callback.ts (hunk 6 above)
Comment (tusharpandey13) [reply src/middleware/callback.ts]: Fixed: removed resumeSilentLogin() from error path entirely. 
Also fixed the underlying null-session bug — hook now receives  the actual session via stateStore.set interception.

File: src/session/HonoCookieHandler.ts (hunk 7 above)
Comment (tusharpandey13) [reply src/session/HonoCookieHandler.ts]: Fixed: getCookie() wrapped in try/catch, returns undefined on decode failure.

File: README.md (hunk 8 above)
Comment (tusharpandey13) [reply README.md:55]: Fixed: code now uses APP_BASE_URL (matches Auth0 SDK convention).

File: src/config/Schema.ts (hunk 10 above)
Comment (tusharpandey13) [reply src/config/Schema.ts:119]: Fixed: added onCallback to ConfigurationSchema with z.custom() validator.

File: src/helpers/session.ts (hunk 11 above)
Comment (tusharpandey13) [reply src/helpers/session.ts]: Fixed: changed to z.string().optional() (matches server-js StateData type).

File: src/lib/Exception.ts (hunk 12 above)
Comment (tusharpandey13) [reply src/lib/Exception.ts]: Fixed: file deleted.

File: src/config/Schema.ts (hunk 13 above)
Comment (tusharpandey13) [reply src/config/Schema.ts]: Fixed: removed top-level `sessionStore` from schema. 
Only `session: { store: myStore }` is the valid API (matches client.ts consumer).

File: src/types/session.ts
Comment (tusharpandey13) [reply src/types/session.ts]: Fixed: type updated to `string | string[]`. First element is active encryption key; remaining keys used for decryption (key rotation).  resolveSecret() helper normalizes to string for server-js.

**Comment by tusharpandey13** [comment]
Fixed the 2 bugs as well
1. `cancelSilentLogin()` and `resumeSilentLogin()` now call `next()` — composable in route chains as expected. Internal usage `inattemptSilentLogin()` uses direct `setCookie()` call instead (doesn't need chain continuation).

2. Root cause confirmed: setCookie writes to response, getCookie reads from request — session invisible on same request.
  Solution: intercept stateStore.set() during `completeInteractiveLogin` to capture `StateData` in memory before it hits the cookie layer. Captured session passed directly to hook (no re-read from cookies). Restore in finally block. Enrichment path also uses captured state.

**Comment by Piyush-85** [comment]
@tusharpandey13 Comments are verified from my end and LGTM.

**Review by Piyush-85 — APPROVED** [review]
